Fixes #8287 Fail CI if Analytics string constants are changed

mrudultora opened the pull request:
## Purpose / Description
Strings used in analytics must not change. So, CI must be failed if changes are made to those.

## Fixes
Fixes #8287 

## Approach
Added a custom annotation `@AnalyticsConstant` and fetched the fields via reflection and checked them with hardcoded fields in the unit test.

## How Has This Been Tested?
Ran the test in Android Studio.

## Learning (optional, can help others)
• Working with Annotations.
• Working with Java Reflection.
• If anyone wants to learn about Parameterized test then this PR would be helpful.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,78 @@
+        for (int i = 0; i < field.length; i++) {
+            if (field[i].isAnnotationPresent(AnalyticsConstant.class)) {
+                try {
+                    assertThat(field[i].get(actions), is(listOfConstantFields.get(i)));
```
This enforces an order, which I don't think is intended

mrudultora replied:
Ya that's true.

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,78 @@
+        listOfConstantFields.add("colordict");
+        listOfConstantFields.add("fora");
+        listOfConstantFields.add("nciku");
+        listOfConstantFields.add("eijiro");
```
I don't think we get a failure if a new constant is added without an entry here

mrudultora replied:
We get an exception **ArrayIndexOutOfBoundsException**.

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,78 @@
+    public void checkAnalyticsString() {
+        UsageAnalytics.Actions actions = new UsageAnalytics.Actions();
+        Field[] field;
+        field = actions.getClass().getDeclaredFields();
```
It might be best to globally check for usages of the annotation. I think I've added a helper method to do this

mrudultora replied:
Could you add that method or helper class here?

david-allison-1 replied:
Appears I was mistaken, apologies

david-allison-1 reviewed (commented):
Thanks! Little refining, then the app is much safer to develop for

mrudultora commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,78 @@
+        for (int i = 0; i < field.length; i++) {
+            if (field[i].isAnnotationPresent(AnalyticsConstant.class)) {
+                try {
+                    assertThat(field[i].get(actions), is(listOfConstantFields.get(i)));
```
So for parameterized test per constant should I fetch each field with directly using a name ?

david-allison-1 replied:
For paramaterized tests: reflect and return the list of strings as the parameter, then check against the array in the test

mikehardy commented:
No pressure - I'm just scanning the queue and labeled this as waiting for feedback which is our label for "PR author has something to do before reviewers scan again"

mrudultora commented:
@david-allison-1 Changes done !

• Added two nested classes (as I wanted one of them to be parameterized and other to be non-parameterized).
• Added a javadoc to explain the need of implemeting in this way.
• Works in every case (if new values added or if the current values changed).

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,124 @@
+
+        @Test
+        public void checkAnalyticsString() {
+            assertEquals("There is no such string in Actions class, thus returning null.", analyticsString, getStringFromReflection(analyticsString));
```
What does the message mean to a developer, how are they meant to fix this?

What are the failure cases
What does a dev need to do in each case

mrudultora replied:
This message means that, the string we searched for in the Actions class is not present there.

david-allison-1 replied:
We want the assertion message to give an actionable message to the developer about how they can fix the issue.

If someone adds a constant and sees `"There is no such string in Actions class, thus returning null."` in the test output, then you haven't explained to them how to fix it. For something like this, be as verbose as you need to

david-allison-1 reviewed (commented):
Looks good, just dev documentation to go

mrudultora commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,124 @@
+
+        @Test
+        public void fieldSizeEqualsListOfConstantFields() {
+            assertEquals("Add the new constant here also in the list listOfConstantFields", listOfConstantFields.size(), getFieldSize());
```
Nice question. Did not thought about that. In that case, we have to remove that from list `listOfConstantsField`.

mrudultora replied:
But, would we remove the strings ?

david-allison-1 replied:
We would, but don't leave it to chance

mrudultora replied:
I would add an if-else check. I think this would work.

mrudultora commented:
The test is failing on CI, but passing on local machine. Any reason ?
```
com.ichi2.anki.analytics.AnalyticsConstantsTest > com.ichi2.anki.analytics.AnalyticsConstantsTest$AnalyticsConstantsFieldLengthTest.fieldSizeEqualsListOfConstantFields FAILED
    java.lang.AssertionError: Add the new constant here also in the list listOfConstantFields expected:<28> but was:<29>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at com.ichi2.anki.analytics.AnalyticsConstantsTest$AnalyticsConstantsFieldLengthTest.fieldSizeEqualsListOfConstantFields(AnalyticsConstantsTest.java:108)

```

david-allison-1 commented:
> The test is failing on CI, but passing on local machine. Any reason ?
> 
> ```
> com.ichi2.anki.analytics.AnalyticsConstantsTest > com.ichi2.anki.analytics.AnalyticsConstantsTest$AnalyticsConstantsFieldLengthTest.fieldSizeEqualsListOfConstantFields FAILED
>     java.lang.AssertionError: Add the new constant here also in the list listOfConstantFields expected:<28> but was:<29>
>         at org.junit.Assert.fail(Assert.java:89)
>         at org.junit.Assert.failNotEquals(Assert.java:835)
>         at org.junit.Assert.assertEquals(Assert.java:647)
>         at com.ichi2.anki.analytics.AnalyticsConstantsTest$AnalyticsConstantsFieldLengthTest.fieldSizeEqualsListOfConstantFields(AnalyticsConstantsTest.java:108)
> ```

Unsure - have added to my TODO list, let me know if you figure it out

mrudultora commented:
>  just dev documentation to go

Umm, would you please clarify this?

mrudultora commented:
Here is the image of passing on local machine.

![Untitled](https://user-images.githubusercontent.com/65870389/113928892-045afa00-980d-11eb-9286-3002d3f76b62.png)

david-allison-1 commented:
> > just dev documentation to go

Oh - clarifying the error messages to ensure future developers need to know what to do, and don't do something wrong

mrudultora commented:
Added the dev documentation explaining the meaning of messages shown if tests failed. Also, explained the meaning of Enclosed runner and why it's used here.

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,141 @@
+         * Because whenever a new constant is added in Actions Class but not added to the list present in this
+         * class (listOfConstantFields) the test must fail.
+         */
+        public static int getFieldSize() {
+            UsageAnalytics.Actions actions = new UsageAnalytics.Actions();
+            Field[] field;
+            field = actions.getClass().getDeclaredFields();
+            return field.length;
+        }
```
This is untested (so therefore lazy). I need to get CI working on my repo. If this doesn't work, I'll put in more effort.

```suggestion
        public static long getFieldSize() {
            return Arrays.stream(UsageAnalytics.class.getDeclaredFields())
                    .filter(x -> x.isAnnotationPresent(AnalyticsConstant.class))
                    .count();
        }
```

david-allison-1 replied:
> One thing is missing here. ( `UsageAnalytics.Actions.class` )

🤦 yes, that's what I get for laziness, thanks!

mrudultora commented:
@david-allison-1 Unit Tests passed this time. To be honest, I didn't understand why they passed this time. Please throw some light on this lazy loading part.

david-allison-1 commented:
> @david-allison-1 Unit Tests passed this time. To be honest, I didn't understand why they passed this time. Please throw some light on this lazy loading part.

Lazy (bad) coding - not testing my work

There was a separate hidden field when run under CI.

The code that I wrote filters the list to only values which have the annotation, which removes that hidden field

mrudultora commented:
Thanks for this info ;-)

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:
```diff
@@ -358,46 +358,75 @@ public static boolean isEnabled() {
     }
 
 
-
```
unnecessary spacing

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,141 @@
+        @Test
+        public void fieldSizeEqualsListOfConstantFields() {
+            if (getFieldSize() > listOfConstantFields.size()) {
+                assertEquals("Add the new constant here also in the list listOfConstantFields", listOfConstantFields.size(), getFieldSize());
```
```suggestion
                assertEquals("Add the newly added analytics constant to AnalyticsConstantsTest.listOfConstantFields. NOTE: Constants should not be renamed as we cannot compare these in analytics", listOfConstantFields.size(), getFieldSize());
```

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,141 @@
+            if (getFieldSize() > listOfConstantFields.size()) {
+                assertEquals("Add the new constant here also in the list listOfConstantFields", listOfConstantFields.size(), getFieldSize());
+            } else if (getFieldSize() < listOfConstantFields.size()) {
+                assertEquals("Remove the constant from the list listOfConstantFields", listOfConstantFields.size(), getFieldSize());
```
```suggestion
                assertEquals("If a constant is removed, it should be removed from AnalyticsConstantsTest.listOfConstantFields. NOTE: Constants should not be renamed as we cannot compare these in analytic", listOfConstantFields.size(), getFieldSize());
```

david-allison-1 reviewed (commented):
A few suggestions - left one for you as I don't want to dictate string constants

mrudultora commented:
I have changed the messages (2 of them are same as suggested by David). I think these messages now, would help devs clearly understand the reason of test failure.

david-allison-1 commented on AnkiDroid/src/test/java/com/ichi2/anki/analytics/AnalyticsConstantsTest.java:
```diff
@@ -0,0 +1,144 @@
+         */
+        @Test
+        public void checkAnalyticsString() {
+            assertEquals("There is no such string in Actions class, thus returning null. Re-check if you renamed any string in the analytics string constants of Actions class or AnalyticsConstantsTest.listOfConstantFields. If so, revert them as those string constants must not change as they are compared in analytics.",
```
```suggestion
            assertEquals("Re-check if you renamed any string in the analytics string constants of Actions class or AnalyticsConstantsTest.listOfConstantFields. If so, revert them as those string constants must not change as they are compared in analytics.",
```

david-allison-1 reviewed (commented):
One nit, then looks good to go

david-allison-1 reviewed (approved):
LGTM, thanks!

mikehardy reviewed (approved):
Nice! That should keep analytics comparisons nicely within the lines, thank you
